container/go/pkg/compat/reader.go

```diff
@@ -16,69 +16,314 @@
+// defining the parts of an image:
+// config is the path to the image config.
+// baseManifest is the path to the manifest of the first image in the chain
+// of images.
+// imgTarball is the path to the image tarball. This will be used as the base
+// image if one or more layers are specified.
+// layers are strings where each item has the format val1,val2,val3,val4 where:
+//   val1 is the compressed layer tarball.
+//   val2 is the uncompressed layer tarball.
+//   val3 is the digest file.
+//   val4 is the diffID file.
+// to an ImageParts object.
+func ImagePartsFromArgs(config, baseManifest, imgTarball string, layers []string) (ImageParts, error) {
+	if config == "" {
+		return ImageParts{}, errors.Errorf("image config was not provided")
+	}
+	result := ImageParts{Config: config, BaseManifest: baseManifest, ImageTarball: imgTarball}
+	for _, l := range layers {
+		lp, err := LayerPartsFromString(l)
+		if err != nil {
+			return ImageParts{}, errors.Wrapf(err, "unable to extract layer parts from %q", l)
+		}
+		result.Layers = append(result.Layers, lp)
+	}
+	return result, nil
 }
 
-// ReadWithBaseTarball returns a Image object with tarball at tarballPath as base and layers appended from layersPath.
-func ReadWithBaseTarball(tarballPath string, layersPath []string) (v1.Image, error) {
-	img, err := tarball.ImageFromPath(tarballPath, nil)
+// reader maintains the state necessary to build a legacyImage object from an
+// ImageParts object.
+type reader struct {
+	// parts is the ImageParts being loaded.
+	parts ImageParts
+	// baseManifest is the manifest of the very first base image in the chain
+	// of images being loaded.
+	baseManifest *v1.Manifest
+	// config is the config of the image being loaded.
+	config *v1.ConfigFile
+	// layerLookup is a map from the diffID of a layer to the layer
+	// itself.
+	layerLookup map[v1.Hash]v1.Layer
+}
+
+// loadMetadata loads the image metadata for the image parts in the given
+// reader.
+func (r *reader) loadMetadata() error {
+	cf, err := os.Open(r.parts.Config)
+	if err != nil {
+		return errors.Wrapf(err, "unable to open image config file %s", r.parts.Config)
+	}
+	c, err := v1.ParseConfigFile(cf)
+	if err != nil {
+		return errors.Wrapf(err, "unable to parse image config from %s", r.parts.Config)
+	}
+	r.config = c
+	if r.parts.BaseManifest == "" {
+		// Base manifest is optional. It's only needed for images whose base
+		// manifests have foreign layers.
+		return nil
+	}
+	mf, err := os.Open(r.parts.BaseManifest)
 	if err != nil {
-		return nil, errors.Wrapf(err, "unable to parse image from tarball at %s", tarballPath)
+		return errors.Wrapf(err, "unable to open base image manifest file %s", r.parts.BaseManifest)
+	}
+	m, err := v1.ParseManifest(mf)
+	if err != nil {
+		return errors.Wrapf(err, "unable to parse base image manifest from %s", r.parts.BaseManifest)
+	}
+	r.baseManifest = m
+	return nil
+}
+
+// foreignLayer represents a foreign layer usually present in windows images.
```

Comment:
Capitalize Windows here and in other comments

Reply:
Done